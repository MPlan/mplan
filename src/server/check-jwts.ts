import { Request, Response, NextFunction } from 'express';
import * as HttpStatus from 'http-status';
import axios from 'axios';
import { getOrThrow } from '../utilities/utilities';
const jwkToPem = require('jwk-to-pem');
import * as jwt from 'jsonwebtoken';
import { AccessTokenPayload } from 'models/token';

const jwkUri = getOrThrow(process.env.JWK_URI);
const audience = getOrThrow(process.env.CLIENT_ID);
const issuer = getOrThrow(process.env.ISSUER);

let publicKeyPem = undefined as undefined | string;
export async function getPublicKeyFromJwk() {
  if (publicKeyPem) return publicKeyPem;

  const response = await axios.get(jwkUri);
  const jwk = response.data.keys[0];
  const pem = jwkToPem(jwk) as string;
  publicKeyPem = pem;
  return pem;
}

export async function checkJwts(req: Request, res: Response, next: NextFunction) {
  if (process.env.NODE_ENV !== 'production') {
    req.user = {
      nickname: process.env.TEST_USERNAME
    };
    next();
    return;
  }

  const authorization = req.headers.authorization;
  if (!authorization) {
    res.sendStatus(HttpStatus.UNAUTHORIZED);
    return;
  }
  const tokenMatch = /Bearer\s*(.*)/.exec(authorization);
  if (!tokenMatch) {
    res.sendStatus(HttpStatus.UNAUTHORIZED);
    return;
  }
  const token = tokenMatch[1];
  if (!token) {
    res.sendStatus(HttpStatus.UNAUTHORIZED);
    return;
  }

  try {
    const pem = await getPublicKeyFromJwk();
    const result = jwt.verify(token, pem) as AccessTokenPayload;

    req.user = result;
    req.user.nickname = result.sub;
    next();
  } catch (e) {
    console.warn(e);
    res.sendStatus(HttpStatus.UNAUTHORIZED);
    return;
  }
}
