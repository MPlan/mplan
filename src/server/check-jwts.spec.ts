require('dotenv').config();
import { getPublicKeyFromJwk } from './check-jwts';

describe('getPublicKeyFromJwk', () => {
  it('gets the public key from the link', async () => {
    const pem = await getPublicKeyFromJwk();
    expect(pem).toBeDefined();
    expect(pem).toMatchSnapshot();
  });
});
