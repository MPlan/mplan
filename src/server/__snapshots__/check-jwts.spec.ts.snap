// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPublicKeyFromJwk gets the public key from the link 1`] = `
"-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAqt6yOiI/wCoCVlGO0MyS
sez0VkSqhPvDl3rfabOslx35mYEO+n4ABfIT5Gn2zN+CeIcOZ5ugAXvIIRWv5H55
+tzjFazi5IKkOIMCiz5//MtsdxKCqGlZu2zt+BLpqTOAPiflNPpM3RUAlxKAhnYE
qNha6+allPnFQupnW/eTYoyuzuedT7dSp90ry0ZcQDimntXWeaSbrYKCj9Rr9W1j
n2uTowUuXaScKXTCjAmJVnsD75JNzQfa8DweklTyWQF+Y5Ky039I0VIu+0CIGhXY
48GAFe2EFb8VpNhf07DP63p138RWQ1d3KPEM9mYJVpQC68j3wzDQYSljpLf9by7T
GwIDAQAB
-----END PUBLIC KEY-----
"
`;
